blockmanager.go

```diff
@@ -319,6 +324,492 @@ func (b *blockManager) handleDonePeerMsg(peers *list.List, sp *ServerPeer) {
 	}
 }
 
+// cfHandler is the cfheader download handler for the block manager. It must be
+// run as a goroutine. It requests and processes cfheaders messages in a
+// separate goroutine from the peer handlers.
+func (b *blockManager) cfHandler() {
+	// If a loop ends with a quit, we want to signal that the goroutine is
+	// done.
+	defer func() {
+		log.Trace("Committed filter header handler done")
+		b.wg.Done()
+	}()
+
+	// Wait for block header sync to complete first.
+	select {
+	case <-b.startCFHeaderSync:
+	case <-b.quit:
+		return
+	}
+
+	// Query all peers for their checkpoints.
+	lastHeader, _, err := b.server.BlockHeaders.ChainTip()
+	if err != nil {
+		log.Critical(err)
+		return
+	}
+	lastHash := lastHeader.BlockHash()
+
+	var regCheckpoints, extCheckpoints []*chainhash.Hash
+
+	for regCheckpoints == nil {
+		select {
+		case <-b.quit:
+			return
+		}
+		allRegCheckpoints := b.getCheckpts(&lastHash,
+			wire.GCSFilterRegular)
+		if len(allRegCheckpoints) == 0 {
+			time.Sleep(QueryTimeout)
+			continue
+		}
+		regCheckpoints, err = b.resolveConflict(allRegCheckpoints,
+			b.server.RegFilterHeaders, wire.GCSFilterRegular)
+		if err != nil {
+			log.Debugf("got error attempting to determine correct "+
+				"cfheader checkpoints: %v, trying again", err)
+			time.Sleep(QueryTimeout)
+		}
+	}
+
+	for extCheckpoints == nil {
+		select {
+		case <-b.quit:
+			return
+		}
+		allExtCheckpoints := b.getCheckpts(&lastHash,
+			wire.GCSFilterExtended)
+		if len(allExtCheckpoints) == 0 {
+			time.Sleep(QueryTimeout)
+			continue
+		}
+		extCheckpoints, err = b.resolveConflict(allExtCheckpoints,
+			b.server.ExtFilterHeaders, wire.GCSFilterExtended)
+		if err != nil {
+			log.Debugf("got error attempting to determine correct "+
+				"cfheader checkpoints: %v, trying again", err)
+			time.Sleep(QueryTimeout)
+		}
+	}
+
+	// At this point, we have good checkpoints and we've banned most peers
+	// that aren't on the same chain as we are. Now, we download all of
+	// the headers that match those checkpoints. We launch the downloads
+	// for both types of headers simultaneously and then wait for them to
+	// both return. Each one returns when its store is caught up with
+	// the block headers.
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		b.getAllCFHeaders(regCheckpoints, b.server.RegFilterHeaders,
+			wire.GCSFilterRegular)
+		wg.Done()
+	}()
+
+	go func() {
```

Comment:
In my branch, I've just deleted all references. But yeah, it would make sense to instead just have everything layered as a map, so future changes don't require that much modification.